osu.Game/Rulesets/Edit/HitObjectComposer.cs

```diff
@@ -293,7 +293,13 @@ public override double GetSnappedDurationFromDistance(double referenceTime, floa
 
         public override float GetSnappedDistanceFromDistance(double referenceTime, float distance)
         {
-            var snappedEndTime = BeatSnapProvider.SnapTime(referenceTime + DistanceToDuration(referenceTime, distance), referenceTime);
+            double actualDuration = referenceTime + DistanceToDuration(referenceTime, distance);
+
+            double snappedEndTime = BeatSnapProvider.SnapTime(actualDuration, referenceTime);
+
+            // we don't want to exceed the actual duration and snap to a point in the future.
+            if (snappedEndTime > actualDuration)
```

Comment:
Is this potentially going to lead to precision issues? Something like:
```
actualDuration = 1.73
snappedEndTime = 1.7330000163
```
And then it'd go back one beatlength and under-snap? Maybe this should be written like `snappedEndTime > actualDuration + beatLength / 2`?

Reply:
In practice, this is only ever invoked on user interaction, so even if it was a precision issue the cursor movement would cancel it out.

We could definitely change to your logic with no issue, but i'm worried it doesn't read as clear...

Reply:
At the very least I think using `Precision.DefinitelyBigger` would be in good form.

Reply:
Ah, never mind, I see what you mean. It has the potential for catastrophic failure.

Reply:
div2 can't work because the input distance is not necessarily snapped. how about using `Precision.AlmostGreater` with 1ms or something?

Reply:
1ms is also fine in my books, if that works.

Reply:
Just added 1ms in the end. Precision isn't really usable in this case without some very backwards logic.